### PR TITLE
TS: loading Swagger files and support for blaming

### PR DESCRIPTION
**Pull request by olydis**
- loads and merges literate Swagger files specified in `input-file` of the config
- support for retrieving the tree of blame/shame for a given problem

Example shaming tree: Who is responsible for `$.definitions.RetentionPolicy.xml`?
```
node:
  source: swagger/swagger.yaml
  column: 4
  line: 24
  name: 'blameRoot ({"path":["$","definitions","RetentionPolicy","xml"]})'
blaming:
  - node:
      source: 'swagger/deliteralize/file%3A%2F%2F%2FC%3A%2FUsers%2Fjobader%2FDesktop%2Fasd%2Fswagger.md'
      line: 24
      column: 4
      name: $.definitions.RetentionPolicy.xml
    blaming:
      - node:
          source: 'input/file%3A%2F%2F%2FC%3A%2FUsers%2Fjobader%2FDesktop%2Fasd%2Fswagger.md'
          line: 1917
          column: 11
          name: $.definitions.RetentionPolicy.xml
        blaming: []
```

The leaf node reads as: File `file:///C:/Users/jobader/Desktop/asd/swagger.md`, line 1917, column 11, also known as `$.definitions.RetentionPolicy.xml` is the root cause.

### CODE REVIEW

File: src/autorest/lib/configuration/configuration.ts [hunk 1]
```diff
@@ -3,6 +3,41 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
+import * as URI from "urijs";
+import { resolveUri } from "../io/input";
+
 export interface AutoRestConfiguration {
+  "input-file": string[] | string;
```
Comment (fearthecowboy) [inline src/autorest/lib/configuration/configuration.ts]: Hmm... `purpose-noun` didn't seem so silly when I first thought of it...

File: src/autorest/lib/configuration/configuration.ts [hunk 2]
```diff
@@ -3,6 +3,41 @@
...
+    const baseFolder = this.config["base-folder"] || "";
+    const baseFolderUri = resolveUri(configFileFolderUri, baseFolder);
+    return baseFolderUri.replace(/\/$/g, "") + "/";
+  }
+
+  private resolveUri(path: string): string {
+    return resolveUri(this.baseFolderUri, path);
+  }
+
+  private inputFiles(): string[] {
```
Comment (fearthecowboy) [inline src/autorest/lib/configuration/configuration.ts]: We're gonna do an awful lot of this type of thing....

File: src/autorest/lib/data-store/dataStore.ts [hunk 3]
```diff
@@ -176,11 +178,12 @@ class DataStoreFileView extends DataStoreView {
 export class DataStore extends DataStoreView {
   private store: Store = {};
 
-  private validate(key: string): void {
+  private async validate(key: string): Promise<void> {
```
Comment (fearthecowboy) [inline src/autorest/lib/data-store/dataStore.ts]: Is `Promise<void>` any different than just `Promise` ?

File: src/autorest/tsconfig.json [hunk 4]
```diff
@@ -1,6 +1,7 @@
 {
   "compilerOptions": {
     "alwaysStrict": true,
+    "forceConsistentCasingInFileNames": true,
```
Comment (fearthecowboy) [inline src/autorest/tsconfig.json]: We should probably try to go to great lengths to ensure that all the tsconfig files are kept as close to exactly the same as possible, given the massive amount of reuse/sharing we're looking at doing.

File: src/autorest/lib/configuration/configuration.ts (hunk 2 above)
Comment (olydis) [inline src/autorest/lib/configuration/configuration.ts]: indeed, gonna abstract "as necessary" :-D

**Pull request merged**